src/psyclone/psyad/tl2ad.py

```diff
@@ -114,31 +110,31 @@ def generate_adjoint_str(tl_fortran_str, active_variables,
     logger.debug("PSyIR after TL preprocessing\n%s",
                  tl_psyir.view(colour=False))
 
-    # TL to AD translation
+    # TL to AD translation and test harness generation
+
+    writer = FortranWriter()
+
     if not api:
         ad_psyir = generate_adjoint(tl_psyir, active_variables)
+        if create_test:
+            test_psyir = generate_adjoint_test(tl_psyir, ad_psyir,
+                                               active_variables)
     elif api == "dynamo0.3":
         ad_psyir = generate_lfric_adjoint(tl_psyir, active_variables)
+        if create_test:
+            test_psyir = generate_lfric_adjoint_test(writer(tl_psyir))
```

Comment:
It feels wrong that we are going back to a string again. We should pass in PSyIR as we do for the other routines.

Reply:
Oh yes, that's true. I think I went this route because I needed an fparser1 parse tree for the metadata but I can encapsulate that within the routine itself.

Reply:
All done with appropriate TODOs to remove the conversion to an fparser1 parse tree.